### PR TITLE
Adds backend widgets content fixing #1307

### LINKED ISSUE #1307: Backend > Widgets needs content
### Plone Documentation versions affected

- [ ] 5.x
- [X] 6.x

### Description

Backend > Widgets needs content.

Source material:

- https://github.com/plone/training/blob/main/docs/mastering-plone/dexterity_reference.md
- https://github.com/plone/plone.app.dexterity/blob/master/docs/reference/widgets.rst

### PR DESCRIPTION
**Pull request by rber474**

### CODE REVIEW

**Comment by rber474** [comment]
Test failed because of the broken link https://52.demo.plone.org in install/index.md reported in issue #1553 and commented  by @stevepiercy days ago in https://community.plone.org/t/5-2-demo-is-down/18059?u=rber474

**Comment by stevepiercy** [comment]
We can ignore the failed link until a decision is made about what to do with that site.

File: docs/backend/widgets.md [hunk 1]
```diff
@@ -12,15 +12,652 @@ myst:
...
+
+        # Note we need to be discreet to different form modes (view, edit, hidden)
+        if self.fields["sections"].mode == INPUT_MODE:
+
+            # Modify a widget with certain name for our purposes
+            widget = self.widgets["sections"]
+
+            # widget.template is a template factory -
+            # Widget.render() will associate later this factory with the widget
+            widget.template = Z3ViewPageTemplateFile("templates/sections.pt")
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: Should this filename match the one give above, `yourwidget.pt`?

File: docs/backend/widgets.md [hunk 2]
```diff
@@ -12,15 +12,652 @@ myst:
...
+`plone.app.z3cform` renders [a frame around each widget](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt), which usually consists of:
+
+-   Label
+-   Required marker
+-   Description
+
+You might want to customize this widget frame for your own form.
+Below is an example of how to do it.
+
+Copy [`widget.pt`](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt) to your own package, and edit it.
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: The subsequent code example names a template as `demo-widget.pt`. Should this line mention that name?

Also what is the file path to the template in the package?

File: docs/backend/widgets.md [hunk 3]
```diff
@@ -12,15 +12,652 @@ myst:
...
+
+    # Add min and max CSS class rendering hints
+    for widget in self.widgets.values():
+        if isinstance(widget, z3c.form.browser.object.ObjectWidget):
+            widget.template = Z3ViewPageTemplateFile("templates/minmax.pt")
+            widget.addClass("min-max-widget")
+            zope.interface.alsoProvides(widget, IFilterWidget)
+```
+
+And then the page template which renders both 0. widget (min) and 1. widget (max) on the same line.
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: I have no idea what this is supposed to mean or say. Please clean this up to clarify the subsequent code example.

**Review by stevepiercy — CHANGES_REQUESTED** [review]
This is excellent. I have a couple of comments and questions about the code examples. Please have a look. Thank you!

File: docs/backend/widgets.md (hunk 3 above)
Comment (rber474) [reply docs/backend/widgets.md]: Would this be better?

`And then create the page template minmax.pt in the same directory of your form module and paste the following code which renders both widgets (min and max) in a single row.`

Comment (stevepiercy) [inline docs/backend/widgets.md]: Ah, now that makes sense. Let's use this.

```suggestion
Then create the page template `minmax.pt` in the same directory as your form module.
Paste the following code in this file.
The code renders both widgets, {guilabel}`min` and {guilabel}`max`, in a single row.
```

File: docs/backend/widgets.md [hunk 4]
```diff
@@ -12,15 +12,652 @@ myst:
...
+
+``` {code-block} html
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: The only reason to use `{codeblock}` is when you want to use Sphinx attributes, such as `emphasize-lines`.

````suggestion
```html
````

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Please remove the three submodules. We only update them when deploying the docs.

File: docs/backend/widgets.md (hunk 3 above)
Comment (rber474) [reply docs/backend/widgets.md]: added suggestion

**Comment by rber474** [comment]
I am not sure if deleting the submodules won't be remove them in the main branch. Please, review that @stevepiercy

File: docs/backend/widgets.md [hunk 5]
```diff
@@ -12,15 +12,688 @@ myst:
...
+To access a widget that's part of a group, you can't use the `updateWidgets()` method.
+The groups and their widgets get initialized after the widgets have been updated.
+Before that, the `groups` attribute is just a list of group factories.
+During the update method though, the groups have been initialized and have their own widget list each.
+To access widgets in a group, you have to access the group in the update method:
+
+``` {code-block} python
+:linenos:
+
+from z3c.form import form
```
Comment (stevepiercy) [inline docs/backend/widgets.md:215]: Why are these added throughout the page? Line numbers don't add anything to these code examples. They just add clutter, and make `{code-block}` and an extra line break unnecessary. If you refer to a specific line number in the narrative text, or emphasize lines, then it makes sense to use the verbose form.

The above three lines (and elsewhere in this page) may be simplified as shown:

````
```python
````

File: docs/backend/widgets.md [hunk 6]
```diff
@@ -12,15 +12,688 @@ myst:
...
+
+
+### `plone.autoform` `widget` directive
+
+`plone.autoform` builds custom `z3c.form` forms based on a model (schema) of fields, and their widgets and options.
+This model is defined as a `zope.schema` based schema, but extra hints can be supplied to control aspects of form display not normally specified in a Zope schema.
+
+By default, `z3c.form` picks a widget based on the type of your field.
+You can change the widget using the `widget` directive if you want users to enter or view data in a different format.
+For example, you can change the widget for the `human` ``boolean`` field to use "yes" and "no" radio buttons instead of its default checkbox:
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: "boolean" is not an inline literal.

Side note, we use only single backticks for inline literals in MyST or Markdown.

```suggestion
For example, you can change the widget for the `human` boolean field to use "yes" and "no" radio buttons instead of its default checkbox:
```

File: docs/backend/widgets.md [hunk 7]
```diff
@@ -12,15 +12,688 @@ myst:
...
+-   Label
+-   Required marker
+-   Description
+
+You might want to customize this widget frame for your own form.
+Below is an example of how to do it.
+
+Copy [`widget.pt`](https://github.com/plone/plone.app.z3cform/blob/master/plone/app/z3cform/templates/widget.pt) to your own package, rename it as ``demo-widget.pt`` and edit it.
+
+Then add the following code to `configure.zcml`. Remember to fix the path of the template according to your own paths.
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: One sentence per line. 

```suggestion
Then add the following code to `configure.zcml`.
Remember to fix the path of the template according to your own paths.
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]

**Comment by stevepiercy** [comment]
> I am not sure if deleting the submodules won't be remove them in the main branch. Please, review that @stevepiercy

I meant to remove the commits that updated the submodule tips. Sorry for the confusion. I think because this PR comes from your fork, things are different from what I understand. Go ahead and restore the submodule tips to the latest from `documentation/6.0`. Thank you!

File: docs/backend/widgets.md (hunk 5 above)
Comment (rber474) [reply docs/backend/widgets.md:215]: didn't notice your commits. fixed again

**Comment by rber474** [comment]
@stevepiercy Thanks for your patient! I think everything is fixed now.

File: docs/backend/widgets.md [hunk 8]
```diff
@@ -12,15 +12,677 @@ myst:
...
+-   Button Widget
+-   Submit Widget
+-   Date Widget
+-   DateTime Widget
+-   Time Widget
+-   DateTime Picker
+
+## Changing a field's widget
+
+You can change a the widget that you use for a field in several ways.
```
Comment (stevepiercy) [inline docs/backend/widgets.md]:
```suggestion
You can change the widget that you use for a field in several ways.
```

File: docs/backend/widgets.md [hunk 9]
```diff
@@ -12,15 +12,677 @@ myst:
...
+                           maxlength view/maxlength;
+                           placeholder view/placeholder;
+                           autocapitalize view/autocapitalize;" />
+</html>
+```
+
+Now you can override the template factory in the `updateWidgets()` method of your form class.
+
+```python
+
```
Comment (stevepiercy) [inline docs/backend/widgets.md]:
```suggestion
```

File: docs/backend/widgets.md [hunk 10]
```diff
@@ -12,15 +12,677 @@ myst:
...
+        permission="zope.Public"
+        template="path/to/template/demo-widget.pt"
+        />
+
+```
+
+Then create a new marker interface in Python code.
+
+```python
+
```
Comment (stevepiercy) [inline docs/backend/widgets.md]:
```suggestion
```

File: docs/backend/widgets.md [hunk 11]
```diff
@@ -12,15 +12,677 @@ myst:
...
+
+class IDemoWidget(Interface):
+    pass
+```
+
+Then apply this marker interface to your widgets in `form.update()`.
+
+```python
+
```
Comment (stevepiercy) [inline docs/backend/widgets.md]:
```suggestion
```

File: docs/backend/widgets.md [hunk 12]
```diff
@@ -12,15 +12,677 @@ myst:
...
+
+## Combined widgets
+
+You can combine several widgets into one with `z3c.form.browser.multil.MultiWidget` and `z3c.form.browser.object.ObjectWidget` classes.
+
+The following example shows how to create an input widget with minimum and maximum values.
+
+```python
+
```
Comment (stevepiercy) [inline docs/backend/widgets.md]:
```suggestion
```

File: docs/backend/widgets.md [hunk 13]
```diff
@@ -12,15 +12,677 @@ myst:
...
+    my_combined_field = zope.schema.Object(
+        __name__="minmax",
+        title=label,
+        schema=IMinMax,
+        required=False
+    )
+
+```
+
+Then you set the {guilabel}`my_combined_field` widget template in `updateWidgets()`:
```
Comment (stevepiercy) [inline docs/backend/widgets.md]: I made a mistake. This is just an inline literal, not a GUI element.

```suggestion
Then you set the `my_combined_field` widget template in `updateWidgets()`:
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Just a little tidying up and it's good to go. Thank you!

**Comment by stevepiercy** [comment]
OK, I cleaned up some extra white spaces and updated the submodule tips to the current ones, instead of deleting them, so that they matched those in `6.0` and dropped off this PR. Let's see what the CI says.

**Review by stevepiercy — APPROVED** [review]

**Comment by stevepiercy** [comment]
The two broken links are known. I notified evee of their bad secure certificate yesterday, and we are waiting for someone to either bring up or take down the 5.2 demo.

**Comment by stevepiercy** [comment]
Thank you!
